File: trie/trie.php
<?php
class TrieNode {
    public $children = [];
    public $isEndOfWord = false;
}

class Trie {
    private $root;
    
    public function __construct() {
        $this->root = new TrieNode();
    }
    
    public function insert($word) {
        $current = $this->root;
        
        for ($i = 0; $i < strlen($word); $i++) {
            $char = $word[$i];
            if (!isset($current->children[$char])) {
                $current->children[$char] = new TrieNode();
            }
            $current = $current->children[$char];
        }
        
        $current->isEndOfWord = true;
    }
    
    public function search($word) {
        $current = $this->root;
        
        for ($i = 0; $i < strlen($word); $i++) {
            $char = $word[$i];
            if (!isset($current->children[$char])) {
                return false;
            }
            $current = $current->children[$char];
        }
        
        return $current->isEndOfWord;
    }
}

$trie = new Trie();
$trie->insert("hello");
echo $trie->search("hello") ? "Found" : "Not found";
?>

File: trie/trie.clj
(defn make-trie-node [] {:children {} :is-end false})

(defn insert-word [trie word]
  (reduce (fn [node char]
           (update-in node [:children char] #(or % (make-trie-node))))
         trie word)
  (assoc-in trie (concat [:children] (interleave word (repeat :children)) [:is-end]) true))

(defn search-word [trie word]
  (let [path (concat [:children] (interleave word (repeat :children)) [:is-end])]
    (get-in trie path false)))

(def trie (-> (make-trie-node)
              (insert-word "hello")))
(println (search-word trie "hello"))